components/ht16k33/ht16k33.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2022 Timofei Korostelev <timofei_public@dranik.dev>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+#include <memory.h>
+
+#include <freertos/FreeRTOS.h>
+
+#include <esp_check.h>
+#include <esp_err.h>
+#include <esp_log.h>
+
+#include "ht16k33.h"
+
+static const char* TAG = "ht16k33";
+
+// All commands are one byte long.
+// First 4 bits are command type.
+// Last 4 bits are parameters.
+#define HT16K33_CMD_RAM_SET_POINTER 0b0000
+#define HT16K33_CMD_SYSTEM_SETUP 0b0010
+#define HT16K33_CMD_SET_BRIGHTNESS 0b1110
+#define HT16K33_CMD_DISPLAY_SETUP 0b1000
+#define HT16K33_CMD_ROW_INT_SET 0b1010
+
+#define HT16K33_SET_RAM_CMD_SIZE_BYTES HT16K33_RAM_SIZE_BYTES + 1
+
+esp_err_t zero_ram(i2c_dev_t *dev) {
+    uint8_t all_zeros[HT16K33_RAM_SIZE_BYTES];
+    memset(all_zeros, 0, HT16K33_RAM_SIZE_BYTES);
+
+    return ht16k33_ram_write(dev, all_zeros);
+}
+
+// Send one-byte command.
+esp_err_t send_short_cmd(i2c_dev_t *dev, uint8_t cmd) {
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_write(dev, NULL, 0, &cmd, sizeof(uint8_t)));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+esp_err_t ht16k33_init(i2c_dev_t *dev, i2c_port_t port, uint32_t i2c_freq_hz,
```

Comment:
1. Descriptor initializer `..._init_desc()` must be decalred separately from device initializer `..._init()`. This is the standard that all i2c device drivers in the library adhere to.
2. In the descriptor initializer itself, you must first check whether the passed pointer to the descriptor is NULL.
3. The descriptor initializer must not accept I2C frequency as a parameter (you cannot set I2C frequency for ESP8266), it must automatically set the highest frequency supported by the device or by the ESP32 itself (1 MHz).
4. The descriptor initializer must not write zeros to the descriptor. This will override any specific settings the user may want to have (see FAQ, question about using pull-up resistors).

Reply:
That was fast.  
Thank you, I will fix it.

Reply:
Thus, you need to split this function into two: the descriptor initializer (taking into account the comments above) and the initializer of the device itself.

Reply:
Yeah, on the I2C frequency.
I don't support automatically setting it, because sometimes it matters.
As an example, HT16K33 and DS3231 support up to 400 kHz as stated in the datasheet.
In HT16K33, the brightness PWM frequency actually depends on I2C clock.
Also, I'm not intending this driver to be compatible with ESP8266, because its legacy. There already is ESP32-C3 as a replacement.
Only ESP32 is specified in the driver manifest.

Reply:
> I don't support automatically setting it, because sometimes it matters.
...
> Also, I'm not intending this driver to be compatible with ESP8266, because its legacy. 

Support for ESP8266 for I2C drivers is not worth much, so it's better to stick to the policy of maximum compatibility in the future.
I'm sorry, but I can't accept the PR without these edits: it would violate the fundamental principle of maximum ESP8266 support and the uniformity of driver interfaces in this library.

Reply:
Of course, if someone will want to add a ESP8266 compatibility, I won't mind it. I just don't like to spend time on it, as it a little bit different.

Reply:
Ok, lets find a compromise.
I will add another initialization function with frequency setting, because I actually need it.